AsyncDisplayKit/ASVideoNode.mm

```diff
@@ -0,0 +1,276 @@
+
+
+#import "ASVideoNode.h"
+
+@interface ASVideoNode () {
+  ASDN::RecursiveMutex _lock;
+  
+  __weak id<ASVideoNodeDatasource> _datasource;
+  
+  BOOL _shouldBePlaying;
+  AVAsset *_asset;
+  AVPlayerItem *_currentItem;
+  ASButtonNode *_playButton;
+  ASDisplayNode *_playerNode;
+  ASDisplayNode *_spinner;
+}
+
+@end
+
+@interface ASDisplayNode ()
+- (void)setInterfaceState:(ASInterfaceState)newState;
+@end
+
+@implementation ASVideoNode
+
+- (instancetype)init {
+  if (!(self = [super init])) { return nil; }
+  
+  _playerNode = [[ASDisplayNode alloc] initWithLayerBlock:^CALayer *{
+    AVPlayerLayer *playerLayer = [[AVPlayerLayer alloc] init];
+    playerLayer.player = [[AVPlayer alloc] init];
+    return playerLayer;
+  }];
+  [self addSubnode:_playerNode];
+  
+  self.gravity = ASVideoGravityResizeAspect;
+  
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(restartVideo:) name:AVPlayerItemDidPlayToEndTimeNotification object:nil];
+  
+  return self;
+}
+
+- (void)setInterfaceState:(ASInterfaceState)newState
+{
+  [super setInterfaceState:newState];
+  
+  if (!(newState & ASInterfaceStateVisible)) {
+    [self pause];
+    [_spinner removeFromSupernode];
+  } else {
+    if (_shouldBePlaying) {
+      [self play];
+    }
+    if (_spinner) {
```

Comment:
Should we have to check here to add the spinner as a subnode?  Maybe we should call -startSpinnerIfNeeded or something, but just adding it seems weird.